### DISCUSSION TITLE
Automating development of pipeline nf-tests

### DISCUSSION THREAD

**Discussion by nschcolnicov** [askimed/nf-test#250] [Ideas]
I recently had to develop nf-tests in bulk for nf-core/smrnaseq, and developed a series of scripts to achieve this in a semi-automated way. There are many improvements to be done, in terms of optimization and further automation, but the scripts were able to succesfully generate working nf-tests which are now part of the dev branch (https://github.com/nf-core/smrnaseq/pull/381).

There is already a debate on whether if we should validate the process output channels or the actual files published in the outdir, my approach does the latter.

Would love to hear your thoughts about adding a similar feature to the nf-test tools.

https://github.com/nschcolnicov/nf-test-helper

@lukfor @sateeshperi @seppinho @edmundmiller

**Comment by maxulysse** [comment]
Hi,
I've going down the same road but for nf-core/rnaseq (https://github.com/nf-core/rnaseq/pull/1360).
I went through a similar process but with less python and less automation for now.
I'm also wondering if we should be checking that the files are published which is what I think we're currently doing with the pipeline level tests (plus the occasional md5sum), but I think it would make more sense to emit channels out or the pipeline and tests that similarly to what we do at the module/subworkflow level (but currently happy with what we're doing).
Main struggle I had was hitting the more than 254 item in a collection to snapshot, which let me to split the different files by tools, and I guess that's exactly why @nschcolnicov came up with the same idea.
It's a bit cumbersome to relaunch the test at the pipeline level since you cannot resume, so I'm also wondering if it could be an option to help out for faster prototyping.

**Comment by sateeshperi** [comment]
hi @lukfor 

@maxulysse @pinin4fjords recently came up with these groovy functions:
[UTILS.groovy](https://github.com/nf-core/rnaseq/pull/1379/files#diff-72c933a46d2f237989238c790c85ee1c9acf1e3eacf06850f256dca2c9b6782d)
[UTILS_PIPELINE.groovy](https://github.com/nf-core/rnaseq/pull/1379/files#diff-5f416edf70907256254727dd0daa6a2e3cc3142f959053688b3323a9bed403ec)

 to semi-automate the capture of stable and unstable elements in snapshots for pipeline level. 
 
 This simplifies a test much further ([example](https://github.com/nf-core/rnaseq/pull/1379/files#diff-a4c7cd455e0e0e8d30c9c22a218c9c70c56f8e717be03c459a3b8d34c4becc77))
 

Would be curious to hear your thoughts and if you see any room for improvement and integration ?

**Reply by maxulysse** [reply]
Hi @lukfor,
So as @sateeshperi, just said, I'm currently playing with @pinin4fjords ideas, and I have been experimenting with pipeline level tests.
My first attempts were clunky

And after some discussions and back and forth, we have a nice for simplifying pipeline level tests.

While I do think being able to emit channels out of the pipeline and capture that similarly to how we do it at the modules and subworkflows level would be amazing, that's not currently what people do with their pipelines, unfortunately.

We think that having a `.nft-ignore` file that can works with regular globs would simplify the usage and syntax of which files one would want to exclude, and even render void the very existence of this `UTILS_PIPELINE.groovy` file.

Thinking something similar to `.gitignore` as you probably guessed with the name.

I think the `UTILS.groovy` can at some point possibly be moved over in a plugin, but that's two simple enough functions for now, so we can wait for it to mature a bit much.

And that would be wonderful

**Reply by nschcolnicov** [reply]
@maxulysse @sateeshperi I really like this approach, how does it look when an assertion for a specific file fails?

**Reply by maxulysse** [reply]
![Screenshot from 2024-09-24 14-05-15](https://github.com/user-attachments/assets/dabb95b3-0317-4b18-a4a7-750741270478)

**Reply by maxulysse** [reply]
I edited the snapshot to rename an expected file and modify an expected md5sum.
it looks nice and easy

**Reply by lukfor** [reply]
Hey! That looks really good! Putting this all in a plugin would be wonderful! 🚀

**Reply by maxulysse** [reply]
Would you say that the nft-ignore idea should be in the plugin?

**Reply by lukfor** [reply]
My first feeling would be yes. Then the 'pipeline-snapshots' plugin would bring the functionality and the configuration.

**Reply by maxulysse** [reply]
ok, then I'll try to include all that nicely in https://github.com/nf-core/nft-utils/

**Reply by lukfor** [reply]
cool! Thanks 🙏 Looking forwards to see you all in Barcelona

**Reply by maxulysse** [reply]
So as you probably already noticed with
https://github.com/askimed/nf-test-plugins/pull/9
I have a first release of https://github.com/nf-core/nft-utils/
There is obviously room for improvements as I haven't written pure JAVA in such a long time, but at least it's not Python :laughing:.
